cogboard-app/src/main/kotlin/com/cognifide/cogboard/widget/type/ZabbixWidget.kt

```diff
@@ -0,0 +1,128 @@
+        }
+    }
+
+    private fun fetchAuthorizationToken() {
+        val body = attachCredentialsToBody()
+        httpPost(publicUrl, body)
+    }
+
+    private fun attachCredentialsToBody(): JsonObject {
+        val credentials = JsonObject()
+                .put(CogboardConstants.PROP_USER, user)
+                .put(CogboardConstants.PROP_PASSWORD, password)
+        return prepareRequestBody(USER_LOGIN, credentials)
+    }
+
+    override fun handleResponse(responseBody: JsonObject) {
+        val body = responseBody.getString("body")
+        val value = JsonObject(body)
+        if (value.getValue(RESULT).toString().contains("key_")) {
+            send(JsonObject().put("lastValue", extractLastValue(value)))
+        } else if (authorizationToken.containsKey(publicUrl) && !value.getString(RESULT).contains(selectedMetric)) {
+            fetchItemData()
+        } else if (JsonObject(body).containsKey(ERROR)) {
+            sendAuthorizationError(JsonObject(body))
+            logger.error(responseBody.getValue(ERROR))
+        } else {
+            saveToken(value.getString(RESULT))
+            fetchItemData()
+        }
+    }
+
+    private fun extractLastValue(value: JsonObject) =
+            ((value.getValue(RESULT) as JsonArray).list[0] as LinkedHashMap<*, *>)["lastvalue"]
+
+    private fun fetchItemData() {
+        val bodyToSend = prepareRequestBody(ITEM_GET, prepareParams(), authorizationToken[publicUrl])
+        httpPost(publicUrl, bodyToSend)
+    }
+
+    private fun prepareRequestBody(method: String, params: JsonObject, auth: String? = null): JsonObject {
+        val authVal = if (auth == null) {
+            null
+        } else {
+            "\"$auth\""
+        }
+        val jsonObject = JsonObject(
+                """
+                {
+                    "jsonrpc": "2.0",
+                    "method": "$method",
+                    "params": $params,
+                    "id": 1,
+                    "auth": $authVal
+                }
+            """
+        )
+
+        return jsonObject
+    }
+
+    private fun prepareParams(): JsonObject {
+        val jsonObject = JsonObject(
+                """
+                {
+                    "filter": {
+                        "host": [
+                            "$host"
+                        ]
+                    },
+                    "search": {
+                        "key_": "$selectedMetric"
+                    },
+                    "sortfield": "name"
+                }
+            """
+        )
+        return jsonObject
+    }
+
+    private fun saveToken(token: String) {
+        authorizationToken[publicUrl] = token
+    }
+
+    private fun sendAuthorizationError(responseBody: JsonObject) {
+        val error = JsonObject(responseBody.getValue(ERROR).toString())
+        val errorMessage = error.getString("message")
+        val errorCause = error.getString("cause")
+
+        send(JsonObject().put(
```

Comment:
we do not longer need to make this: `JsonObject().put(CogboardConstants.PROP_CONTENT,  JsonObject().....))`
it is now done for us inside `send(..)` function

Reply:
done